src/net/dns.rs

```diff
@@ -42,33 +82,107 @@ pub(crate) async fn prune_dns_cache(context: &Context) -> Result<()> {
     Ok(())
 }
 
-/// Looks up the hostname and updates DNS cache
-/// on success.
+/// Map from hostname to IP addresses.
+///
+/// NOTE: sync mutex is used, so it must not be held across `.await`
+/// to avoid deadlocks.
```

Comment:
Just a thought (i.e. non-blocking): I'm wondering if we should link https://docs.rs/tokio/latest/tokio/sync/struct.Mutex.html#which-kind-of-mutex-should-you-use for people reading the code who have no idea what we're talking about